Feedback thread for experimental spell checker service

[v0.3.13-beta05](https://github.com/florisboard/florisboard/releases/tag/v0.3.13-beta05) adds an experimental spell checker service. This issue thread serves as a collection of feedback for this feature and will be used to improve the spell checker until it releases on the stable channel with v0.3.13. Feel free to comment down below, any feedback from suggestions to bug / crash reports for the spell checker is welcome!

Wiki entry for help with using the spell checker: https://github.com/florisboard/florisboard/wiki/System-spell-checker-service

Revision 1: #1069 (beta06)
Revision 2: #1082 (beta07)
Revision 3: #1082 (beta08)
Revision 4: #1172 (beta10)

### Known issues / Suggested improvements [will be extended dynamically]

- [x] Dictionaries with languages containing Cyrillic characters (Greek, Bulgarian etc) fail to load (especially those from Firefox and Free Office), possibly encoding issue (Fixed in #1069)
- [x] The APK size is too high, try optimizing (especially ICU data) (Fixed in #1082)
- [x] Language is fixed per session, interferes with "Use keyboard subtypes" when switching the subtype while a session is active
- [x] Czech XPI extension from Mozilla does not conform to expected format causing a crash when importing, fix importer script (Fixed in #1069)
- [x] Bulgarian Mozilla extension license file uses different name, add fix (Fixed in #1069)
- [x] Auto-capitalization pref is incorrectly labeled as "Spell checker" (Fixed in #1069)
- [x] FreeOffice link leads to nowhere when used on German devices (make link statically linked to `en`) (Fixed in #1069)
- [ ] Best correction match should show in Smartbar (inline spellchecking)

...

Crash when opening a czech dictionary from Mozzila.

#### Environment information
- FlorisBoard Version: 0.3.13-beta05 (48)
- Device: Samsung SM-A505FN
- Android: 11 (cn=R sdk=30)

#### Features enabled
```kt
smartbar = true
suggestions = false
suggestions_clipboard = true
suggestions_next_word = true
glide = true
clipboard_internal = true
clipboard_history = false
```

#### Attached stacktrace files
<details>
<summary>1626184897059.stacktrace</summary>

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:612)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
	... 1 more
Caused by: java.io.FileNotFoundException: /data/user/0/dev.patrickgold.florisboard.beta/cache/__temp000__ime_spelling_import_new_dict/cs_CZ/cs_CZ.aff: open failed: ENOENT (No such file or directory)
	at libcore.io.IoBridge.open(IoBridge.java:492)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:236)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:186)
	at dev.patrickgold.florisboard.ime.spelling.SpellingManager.prepareImport-gIAlu-s(SpellingManager.kt:213)
	at dev.patrickgold.florisboard.settings.spelling.ImportDictionaryFragment.importDict$lambda-5(ImportDictionaryFragment.kt:51)
	at dev.patrickgold.florisboard.settings.spelling.ImportDictionaryFragment.lambda$B1W-grqL87TnSfoLsCdIZzvpqNI(Unknown Source:0)
	at dev.patrickgold.florisboard.settings.spelling.-$$Lambda$ImportDictionaryFragment$B1W-grqL87TnSfoLsCdIZzvpqNI.onActivityResult(Unknown Source:4)
	at androidx.activity.result.ActivityResultRegistry$1.onStateChanged(ActivityResultRegistry.java:148)
	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354)
	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:265)
	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:307)
	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:148)
	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:134)
	at androidx.fragment.app.Fragment.performStart(Fragment.java:3013)
	at androidx.fragment.app.FragmentStateManager.start(FragmentStateManager.java:586)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:300)
	at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:112)
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1636)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3112)
	at androidx.fragment.app.FragmentManager.dispatchStart(FragmentManager.java:3063)
	at androidx.fragment.app.FragmentController.dispatchStart(FragmentController.java:262)
	at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:482)
	at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:210)
	at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1435)
	at android.app.Activity.performStart(Activity.java:8222)
	at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3818)
	at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:221)
	at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:246)
	at android.app.ActivityThread.main(ActivityThread.java:8512)
	... 3 more
Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
	at libcore.io.Linux.open(Native Method)
	at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:254)
	at libcore.io.ForwardingOs.open(ForwardingOs.java:166)
	at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8373)
	at libcore.io.IoBridge.open(IoBridge.java:478)
	... 36 more

```
</details>

--- a/app/build.gradle.kts
+++ b/app/build.gradle.kts
@@ -1,5 +1,5 @@
 plugins {
-    id("com.android.application") version "7.0.0"
+    id("com.android.application") version "7.0.1"
     kotlin("android") version "1.5.20"
     kotlin("kapt") version "1.5.20"
     kotlin("plugin.serialization") version "1.5.20"

--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -34,7 +34,8 @@
             android:name="dev.patrickgold.florisboard.FlorisImeService"
             android:label="@string/floris_app_name"
             android:permission="android.permission.BIND_INPUT_METHOD"
-            android:directBootAware="true">
+            android:directBootAware="true"
+            android:exported="true">
             <intent-filter>
                 <action android:name="android.view.InputMethod"/>
             </intent-filter>
@@ -45,7 +46,8 @@
         <service
             android:name="dev.patrickgold.florisboard.FlorisSpellCheckerService"
             android:label="@string/floris_app_name"
-            android:permission="android.permission.BIND_TEXT_SERVICE">
+            android:permission="android.permission.BIND_TEXT_SERVICE"
+            android:exported="true">
             <intent-filter>
                 <action android:name="android.service.textservice.SpellCheckerService"/>
             </intent-filter>
@@ -68,7 +70,8 @@
             android:label="@string/floris_app_name"
             android:launchMode="singleTask"
             android:roundIcon="@mipmap/floris_app_icon_round"
-            android:targetActivity="dev.patrickgold.florisboard.setup.SetupActivity">
+            android:targetActivity="dev.patrickgold.florisboard.setup.SetupActivity"
+            android:exported="true">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN"/>
                 <category android:name="android.intent.category.LAUNCHER"/>

--- a/app/src/main/java/dev/patrickgold/florisboard/FlorisSpellCheckerService.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/FlorisSpellCheckerService.kt
@@ -17,24 +17,26 @@
 package dev.patrickgold.florisboard
 
 import android.service.textservice.SpellCheckerService
-import android.util.LruCache
 import android.view.textservice.SentenceSuggestionsInfo
 import android.view.textservice.SuggestionsInfo
 import android.view.textservice.TextInfo
 import dev.patrickgold.florisboard.common.FlorisLocale
 import dev.patrickgold.florisboard.debug.LogTopic
 import dev.patrickgold.florisboard.debug.flogInfo
-import dev.patrickgold.florisboard.ime.core.Preferences
 import dev.patrickgold.florisboard.ime.core.Subtype
 import dev.patrickgold.florisboard.ime.core.SubtypeManager
 import dev.patrickgold.florisboard.ime.dictionary.DictionaryManager
-import dev.patrickgold.florisboard.ime.spelling.SpellingDict
-import dev.patrickgold.florisboard.ime.spelling.SpellingManager
-import kotlin.contracts.InvocationKind
-import kotlin.contracts.contract
+import dev.patrickgold.florisboard.ime.spelling.SpellingService
+import kotlinx.coroutines.runBlocking
 
 class FlorisSpellCheckerService : SpellCheckerService() {
+    companion object {
+        private const val USE_FLORIS_SUBTYPES_LOCALE: String = "zz"
+    }
+
     private val dictionaryManager get() = DictionaryManager.default()
+    private val spellingService: SpellingService = SpellingService.globalInstance()
+    private val subtypeManager get() = SubtypeManager.default()
 
     override fun onCreate() {
         flogInfo(LogTopic.SPELL_EVENTS)
@@ -55,43 +57,8 @@ class FlorisSpellCheckerService : SpellCheckerService() {
         super.onDestroy()
     }
 
-    private class SuggestionsCache(size: Int) {
-        val suggestionsInfoCache: LruCache<String, SuggestionsInfo> = LruCache(size)
-
-        inline fun getOrGenerate(word: String, generator: (w: String) -> SuggestionsInfo): SuggestionsInfo {
-            contract {
-                callsInPlace(generator, InvocationKind.AT_MOST_ONCE)
-            }
-            val cachedSuggestionsInfo = suggestionsInfoCache.get(word)
-            if (cachedSuggestionsInfo != null) {
-                return cachedSuggestionsInfo
-            }
-            val newSuggestionsInfo = generator(word)
-            suggestionsInfoCache.put(word, newSuggestionsInfo)
-            return newSuggestionsInfo
-        }
-
-        fun clear() {
-            suggestionsInfoCache.evictAll()
-        }
-    }
-
-    private class FlorisSpellCheckerSession : Session() {
-        companion object {
-            private const val USE_FLORIS_SUBTYPES_LOCALE: String = "zz"
-            private const val SUGGESTIONS_MAX_SIZE = 50
-
-            private val EMPTY_STRING_ARRAY: Array<out String> = arrayOf()
-        }
-
-        private val prefs get() = Preferences.default()
-        private val dictionaryManager get() = DictionaryManager.default()
-        private val spellingManager get() = SpellingManager.default()
-        private val subtypeManager get() = SubtypeManager.default()
-
-        private var cachedSpellingDict: SpellingDict? = null
+    private inner class FlorisSpellCheckerSession : Session() {
         private var cachedSpellingLocale: FlorisLocale? = null
-        private val suggestionsCache = SuggestionsCache(SUGGESTIONS_MAX_SIZE)
 
         override fun onCreate() {
             flogInfo(LogTopic.SPELL_EVENTS) { "Session locale: $locale" }
@@ -108,48 +75,20 @@ class FlorisSpellCheckerService : SpellCheckerService() {
 
             if (evaluatedLocale != cachedSpellingLocale) {
                 cachedSpellingLocale = evaluatedLocale
-                cachedSpellingDict = spellingManager.getSpellingDict(evaluatedLocale)
-                suggestionsCache.clear()
-            }
-        }
-
-        private fun emptySuggestionsInfo() = SuggestionsInfo(0, EMPTY_STRING_ARRAY)
-
-        private fun spell(
-            textInfo: TextInfo,
-            suggestionsLimit: Int,
-            spellingDict: SpellingDict,
-            spellingLocale: FlorisLocale
-        ): SuggestionsInfo {
-            val word = textInfo.text ?: return emptySuggestionsInfo()
-
-            return suggestionsCache.getOrGenerate(word) {
-                var isWordOk = false
-                if (prefs.spelling.useUdmEntries) {
-                    isWordOk = dictionaryManager.spell(word, spellingLocale)
-                }
-                return@getOrGenerate if (isWordOk) {
-                    SuggestionsInfo(SuggestionsInfo.RESULT_ATTR_IN_THE_DICTIONARY, EMPTY_STRING_ARRAY)
-                } else {
-                    isWordOk = spellingDict.spell(word)
-                    if (isWordOk) {
-                        SuggestionsInfo(SuggestionsInfo.RESULT_ATTR_IN_THE_DICTIONARY, EMPTY_STRING_ARRAY)
-                    } else {
-                        val suggestions = spellingDict.suggest(word, suggestionsLimit)
-                        SuggestionsInfo(SuggestionsInfo.RESULT_ATTR_LOOKS_LIKE_TYPO, suggestions)
-                    }
-                }
             }
         }
 
         private fun spellMultiple(
+            spellingLocale: FlorisLocale,
             textInfos: Array<out TextInfo>,
-            suggestionsLimit: Int,
-            spellingDict: SpellingDict,
-            spellingLocale: FlorisLocale
-        ): Array<SuggestionsInfo> {
-            return Array(textInfos.size) { n ->
-                spell(textInfos[n], suggestionsLimit, spellingDict, spellingLocale).apply {
+            suggestionsLimit: Int
+        ): Array<SuggestionsInfo> = runBlocking {
+            val retInfos = Array(textInfos.size) { n ->
+                val word = textInfos[n].text ?: ""
+                spellingService.spellAsync(spellingLocale, word, suggestionsLimit)
+            }
+            Array(textInfos.size) { n ->
+                retInfos[n].await().apply {
                     setCookieAndSequence(textInfos[n].cookie, textInfos[n].sequence)
                 }
             }
@@ -158,12 +97,11 @@ class FlorisSpellCheckerService : SpellCheckerService() {
         override fun onGetSuggestions(textInfo: TextInfo?, suggestionsLimit: Int): SuggestionsInfo {
             flogInfo(LogTopic.SPELL_EVENTS) { "text=${textInfo?.text}, limit=$suggestionsLimit" }
 
-            textInfo ?: return emptySuggestionsInfo()
+            textInfo?.text ?: return SpellingService.emptySuggestionsInfo()
             setupSpellingIfNecessary()
-            val spellingDict = cachedSpellingDict ?: return emptySuggestionsInfo()
-            val spellingLocale = cachedSpellingLocale ?: return emptySuggestionsInfo()
+            val spellingLocale = cachedSpellingLocale ?: return SpellingService.emptySuggestionsInfo()
 
-            return spell(textInfo, suggestionsLimit, spellingDict, spellingLocale)
+            return spellingService.spell(spellingLocale, textInfo.text, suggestionsLimit)
         }
 
         override fun onGetSuggestionsMultiple(
@@ -175,10 +113,9 @@ class FlorisSpellCheckerService : SpellCheckerService() {
 
             textInfos ?: return emptyArray()
             setupSpellingIfNecessary()
-            val spellingDict = cachedSpellingDict ?: return emptyArray()
             val spellingLocale = cachedSpellingLocale ?: return emptyArray()
 
-            return spellMultiple(textInfos, suggestionsLimit, spellingDict, spellingLocale)
+            return spellMultiple(spellingLocale, textInfos, suggestionsLimit)
         }
 
         override fun onGetSentenceSuggestionsMultiple(
@@ -195,14 +132,12 @@ class FlorisSpellCheckerService : SpellCheckerService() {
             flogInfo(LogTopic.SPELL_EVENTS)
 
             super.onCancel()
-            suggestionsCache.clear()
         }
 
         override fun onClose() {
             flogInfo(LogTopic.SPELL_EVENTS)
 
             super.onClose()
-            suggestionsCache.clear()
         }
     }
 }

--- a/app/src/main/java/dev/patrickgold/florisboard/FlorisSpellCheckerService.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/FlorisSpellCheckerService.kt
@@ -18,6 +18,7 @@ package dev.patrickgold.florisboard
 
 import android.service.textservice.SpellCheckerService
 import android.util.LruCache
+import android.view.textservice.SentenceSuggestionsInfo
 import android.view.textservice.SuggestionsInfo
 import android.view.textservice.TextInfo
 import dev.patrickgold.florisboard.common.FlorisLocale
@@ -69,6 +70,10 @@ class FlorisSpellCheckerService : SpellCheckerService() {
             suggestionsInfoCache.put(word, newSuggestionsInfo)
             return newSuggestionsInfo
         }
+
+        fun clear() {
+            suggestionsInfoCache.evictAll()
+        }
     }
 
     private class FlorisSpellCheckerSession : Session() {
@@ -85,25 +90,36 @@ class FlorisSpellCheckerService : SpellCheckerService() {
         private val subtypeManager get() = SubtypeManager.default()
 
         private var spellingDict: SpellingDict? = null
-        private lateinit var spellingLocale: FlorisLocale
+        private var spellingLocale: FlorisLocale? = null
         private val suggestionsCache = SuggestionsCache(SUGGESTIONS_MAX_SIZE)
 
         override fun onCreate() {
             flogInfo(LogTopic.SPELL_EVENTS) { "Session locale: $locale" }
 
-            spellingLocale = when (locale) {
+            setupSpellingIfNecessary()
+        }
+
+        private fun setupSpellingIfNecessary() {
+            val evaluatedLocale = when (locale) {
                 null -> Subtype.DEFAULT.locale
                 USE_FLORIS_SUBTYPES_LOCALE -> (subtypeManager.getActiveSubtype() ?: Subtype.DEFAULT).locale
                 else -> FlorisLocale.from(locale)
             }
 
-            spellingDict = spellingManager.getSpellingDict(spellingLocale)
+            if (evaluatedLocale != spellingLocale) {
+                spellingLocale = evaluatedLocale
+                spellingDict = spellingManager.getSpellingDict(evaluatedLocale)
+                suggestionsCache.clear()
+            }
         }
 
         override fun onGetSuggestions(textInfo: TextInfo?, suggestionsLimit: Int): SuggestionsInfo {
-            flogInfo(LogTopic.SPELL_EVENTS) { "text=${textInfo?.text}, limit=$suggestionsLimit"}
+            flogInfo(LogTopic.SPELL_EVENTS) { "text=${textInfo?.text}, limit=$suggestionsLimit" }
+
+            setupSpellingIfNecessary()
 
             val spellingDict = spellingDict ?: return SuggestionsInfo(0, EMPTY_STRING_ARRAY)
+            val spellingLocale = spellingLocale ?: return SuggestionsInfo(0, EMPTY_STRING_ARRAY)
             val word = textInfo?.text ?: return SuggestionsInfo(0, EMPTY_STRING_ARRAY)
 
             return suggestionsCache.getOrGenerate(word) {
@@ -123,6 +139,41 @@ class FlorisSpellCheckerService : SpellCheckerService() {
                     }
                 }
             }
+        }
+
+        override fun onGetSuggestionsMultiple(
+            textInfos: Array<out TextInfo>?,
+            suggestionsLimit: Int,
+            sequentialWords: Boolean
+        ): Array<SuggestionsInfo> {
+            flogInfo(LogTopic.SPELL_EVENTS)
+
+            // TODO: implement custom solution here instead of calling the default implementation
+            return super.onGetSuggestionsMultiple(textInfos, suggestionsLimit, sequentialWords)
+        }
+
+        override fun onGetSentenceSuggestionsMultiple(
+            textInfos: Array<out TextInfo>?,
+            suggestionsLimit: Int
+        ): Array<SentenceSuggestionsInfo> {
+            flogInfo(LogTopic.SPELL_EVENTS)
+
+            // TODO: implement custom solution here instead of calling the default implementation
+            return super.onGetSentenceSuggestionsMultiple(textInfos, suggestionsLimit)
+        }
+
+        override fun onCancel() {
+            flogInfo(LogTopic.SPELL_EVENTS)
+
+            super.onCancel()
+            suggestionsCache.clear()
+        }
+
+        override fun onClose() {
+            flogInfo(LogTopic.SPELL_EVENTS)
+
+            super.onClose()
+            suggestionsCache.clear()
         }
     }
 }

--- a/app/src/main/java/dev/patrickgold/florisboard/FlorisSpellCheckerService.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/FlorisSpellCheckerService.kt
@@ -89,8 +89,8 @@ class FlorisSpellCheckerService : SpellCheckerService() {
         private val spellingManager get() = SpellingManager.default()
         private val subtypeManager get() = SubtypeManager.default()
 
-        private var spellingDict: SpellingDict? = null
-        private var spellingLocale: FlorisLocale? = null
+        private var cachedSpellingDict: SpellingDict? = null
+        private var cachedSpellingLocale: FlorisLocale? = null
         private val suggestionsCache = SuggestionsCache(SUGGESTIONS_MAX_SIZE)
 
         override fun onCreate() {
@@ -106,21 +106,22 @@ class FlorisSpellCheckerService : SpellCheckerService() {
                 else -> FlorisLocale.from(locale)
             }
 
-            if (evaluatedLocale != spellingLocale) {
-                spellingLocale = evaluatedLocale
-                spellingDict = spellingManager.getSpellingDict(evaluatedLocale)
+            if (evaluatedLocale != cachedSpellingLocale) {
+                cachedSpellingLocale = evaluatedLocale
+                cachedSpellingDict = spellingManager.getSpellingDict(evaluatedLocale)
                 suggestionsCache.clear()
             }
         }
 
-        override fun onGetSuggestions(textInfo: TextInfo?, suggestionsLimit: Int): SuggestionsInfo {
-            flogInfo(LogTopic.SPELL_EVENTS) { "text=${textInfo?.text}, limit=$suggestionsLimit" }
+        private fun emptySuggestionsInfo() = SuggestionsInfo(0, EMPTY_STRING_ARRAY)
 
-            setupSpellingIfNecessary()
-
-            val spellingDict = spellingDict ?: return SuggestionsInfo(0, EMPTY_STRING_ARRAY)
-            val spellingLocale = spellingLocale ?: return SuggestionsInfo(0, EMPTY_STRING_ARRAY)
-            val word = textInfo?.text ?: return SuggestionsInfo(0, EMPTY_STRING_ARRAY)
+        private fun spell(
+            textInfo: TextInfo,
+            suggestionsLimit: Int,
+            spellingDict: SpellingDict,
+            spellingLocale: FlorisLocale
+        ): SuggestionsInfo {
+            val word = textInfo.text ?: return emptySuggestionsInfo()
 
             return suggestionsCache.getOrGenerate(word) {
                 var isWordOk = false
@@ -141,6 +142,30 @@ class FlorisSpellCheckerService : SpellCheckerService() {
             }
         }
 
+        private fun spellMultiple(
+            textInfos: Array<out TextInfo>,
+            suggestionsLimit: Int,
+            spellingDict: SpellingDict,
+            spellingLocale: FlorisLocale
+        ): Array<SuggestionsInfo> {
+            return Array(textInfos.size) { n ->
+                spell(textInfos[n], suggestionsLimit, spellingDict, spellingLocale).apply {
+                    setCookieAndSequence(textInfos[n].cookie, textInfos[n].sequence)
+                }
+            }
+        }
+
+        override fun onGetSuggestions(textInfo: TextInfo?, suggestionsLimit: Int): SuggestionsInfo {
+            flogInfo(LogTopic.SPELL_EVENTS) { "text=${textInfo?.text}, limit=$suggestionsLimit" }
+
+            textInfo ?: return emptySuggestionsInfo()
+            setupSpellingIfNecessary()
+            val spellingDict = cachedSpellingDict ?: return emptySuggestionsInfo()
+            val spellingLocale = cachedSpellingLocale ?: return emptySuggestionsInfo()
+
+            return spell(textInfo, suggestionsLimit, spellingDict, spellingLocale)
+        }
+
         override fun onGetSuggestionsMultiple(
             textInfos: Array<out TextInfo>?,
             suggestionsLimit: Int,
@@ -148,8 +173,12 @@ class FlorisSpellCheckerService : SpellCheckerService() {
         ): Array<SuggestionsInfo> {
             flogInfo(LogTopic.SPELL_EVENTS)
 
-            // TODO: implement custom solution here instead of calling the default implementation
-            return super.onGetSuggestionsMultiple(textInfos, suggestionsLimit, sequentialWords)
+            textInfos ?: return emptyArray()
+            setupSpellingIfNecessary()
+            val spellingDict = cachedSpellingDict ?: return emptyArray()
+            val spellingLocale = cachedSpellingLocale ?: return emptyArray()
+
+            return spellMultiple(textInfos, suggestionsLimit, spellingDict, spellingLocale)
         }
 
         override fun onGetSentenceSuggestionsMultiple(

--- a/app/src/main/java/dev/patrickgold/florisboard/ime/spelling/SpellingService.kt
+++ b/app/src/main/java/dev/patrickgold/florisboard/ime/spelling/SpellingService.kt
@@ -1,0 +1,148 @@
+/*
+ * Copyright (C) 2021 Patrick Goldinger
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.patrickgold.florisboard.ime.spelling
+
+import android.util.LruCache
+import android.view.textservice.SuggestionsInfo
+import dev.patrickgold.florisboard.common.FlorisLocale
+import dev.patrickgold.florisboard.ime.core.Preferences
+import dev.patrickgold.florisboard.ime.dictionary.DictionaryManager
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.async
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
+
+class SpellingService {
+    companion object {
+        private const val LRU_CACHE_MAX_LOCALE_ENTRIES = 4
+        private const val LRU_CACHE_MAX_SUGGESTIONS_INFO_ENTRIES = 192
+
+        private val EMPTY_STRING_ARRAY: Array<out String> = arrayOf()
+
+        private var globalInstance: SpellingService? = null
+
+        @Synchronized
+        fun globalInstance(): SpellingService {
+            val instance = globalInstance
+            return if (instance != null) {
+                instance
+            } else {
+                val newInstance = SpellingService()
+                globalInstance = newInstance
+                newInstance
+            }
+        }
+
+        fun emptySuggestionsInfo() =
+            SuggestionsInfo(0, EMPTY_STRING_ARRAY)
+
+        fun dictMatchSuggestionsInfo() =
+            SuggestionsInfo(SuggestionsInfo.RESULT_ATTR_IN_THE_DICTIONARY, EMPTY_STRING_ARRAY)
+
+        fun typoSuggestionsInfo(suggestions: Array<out String>) =
+            SuggestionsInfo(SuggestionsInfo.RESULT_ATTR_LOOKS_LIKE_TYPO, suggestions)
+    }
+
+    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
+    private val prefs get() = Preferences.default()
+    private val dictionaryManager get() = DictionaryManager.default()
+    private val spellingManager get() = SpellingManager.default()
+
+    private val globalCache = LruCache<FlorisLocale, SuggestionsCache>(LRU_CACHE_MAX_LOCALE_ENTRIES)
+    private val globalCacheGuard = Mutex(locked = false)
+
+    private suspend fun getSuggestionsCache(locale: FlorisLocale): SuggestionsCache {
+        return globalCacheGuard.withLock {
+            val suCache = globalCache.get(locale)
+            if (suCache != null) {
+                suCache
+            } else {
+                val spellingDict = spellingManager.getSpellingDict(locale)
+                val newCache = SuggestionsCache(LRU_CACHE_MAX_SUGGESTIONS_INFO_ENTRIES, spellingDict)
+                globalCache.put(locale, newCache)
+                newCache
+            }
+        }
+    }
+
+    fun spell(locale: FlorisLocale, word: String, suggestionsLimit: Int): SuggestionsInfo {
+        return runBlocking { spellAsync(locale, word, suggestionsLimit).await() }
+    }
+
+    suspend fun spellAsync(locale: FlorisLocale, word: String, suggestionsLimit: Int): Deferred<SuggestionsInfo> {
+        val suggestionsCache = getSuggestionsCache(locale)
+        return suggestionsCache.getOrGenerateAsync(word) { spellingDict ->
+            var isWordOk = false
+            if (prefs.spelling.useUdmEntries) {
+                isWordOk = dictionaryManager.spell(word, locale)
+            }
+            return@getOrGenerateAsync if (isWordOk) {
+                dictMatchSuggestionsInfo()
+            } else {
+                isWordOk = spellingDict.spell(word)
+                if (isWordOk) {
+                    dictMatchSuggestionsInfo()
+                } else {
+                    val suggestions = spellingDict.suggest(word, suggestionsLimit)
+                    typoSuggestionsInfo(suggestions)
+                }
+            }
+        }
+    }
+
+    private inner class SuggestionsCache(size: Int, private val spellingDict: SpellingDict?) {
+        private val localCache = LruCache<String, Deferred<SuggestionsInfo>>(size)
+        private val localCacheGuard = Mutex(locked = false)
+
+        suspend inline fun getOrGenerateAsync(
+            word: String,
+            crossinline generator: (dict: SpellingDict) -> SuggestionsInfo
+        ): Deferred<SuggestionsInfo> {
+            contract {
+                callsInPlace(generator, InvocationKind.AT_MOST_ONCE)
+            }
+            return localCacheGuard.withLock {
+                val cachedInfo = localCache.get(word)
+                if (cachedInfo != null) {
+                    cachedInfo
+                } else {
+                    val newInfo = scope.async {
+                        if (spellingDict != null) {
+                            generator(spellingDict)
+                        } else {
+                            emptySuggestionsInfo()
+                        }
+                    }
+                    localCache.put(word, newInfo)
+                    newInfo
+                }
+            }
+        }
+
+        suspend fun clear() {
+            localCacheGuard.withLock {
+                localCache.evictAll()
+            }
+        }
+    }
+}

